crates/prover/src/runner.rs

```diff
@@ -202,85 +299,139 @@ where
+        let start_l1_height = match last_scanned_l1_height {
+            Some(height) => height.0,
             None => get_initial_slot_height::<Da::Spec>(&self.sequencer_client).await,
         };
 
-        let mut l2_height = self.start_height;
-
         let prover_config = self.prover_config.clone().unwrap();
 
-        let pg_client = match prover_config.db_config {
+        let pg_client = match prover_config.clone().db_config {
             Some(db_config) => {
                 info!("Connecting to postgres");
                 Some(PostgresConnector::new(db_config.clone()).await)
             }
             None => None,
         };
 
-        loop {
-            let da_service = &self.da_service;
-
-            let exponential_backoff = ExponentialBackoffBuilder::new()
-                .with_initial_interval(Duration::from_secs(1))
-                .with_max_elapsed_time(Some(Duration::from_secs(5 * 60)))
-                .build();
-            let last_finalized_height = retry_backoff(exponential_backoff.clone(), || async {
-                da_service
-                    .get_last_finalized_block_header()
-                    .await
-                    .map_err(backoff::Error::transient)
-            })
-            .await?
-            .height();
+        // Save the genesis state root to ledger db
+        if self.start_l2_height == 1 {
+            self.ledger_db
+                .set_l2_state_root(0, &self.state_root.clone())?;
+        }
 
-            if l1_height > last_finalized_height {
-                sleep(Duration::from_secs(1)).await;
-                continue;
+        let (l1_tx, mut l1_rx) = mpsc::channel(1);
+        let da_service = self.da_service.clone();
+        let l1_block_cache = self.l1_block_cache.clone();
+        let l1_handle = tokio::spawn(async move {
+            l1_sync(start_l1_height, da_service, l1_tx, l1_block_cache).await;
+        });
+        tokio::pin!(l1_handle);
+
+        let (l2_tx, mut l2_rx) = mpsc::channel(1);
+        let start_l2_height = self.start_l2_height;
+        let sequencer_client = self.sequencer_client.clone();
+        let sync_blocks_count = self.sync_blocks_count;
+        let l2_handle = tokio::spawn(async move {
+            sync_l2::<Da>(start_l2_height, sequencer_client, l2_tx, sync_blocks_count).await;
+        });
+        tokio::pin!(l2_handle);
+        let da_service = self.da_service.clone();
+        let l1_block_cache = self.l1_block_cache.clone();
+        let mut pending_l1_blocks: VecDeque<<Da as DaService>::FilteredBlock> =
+            VecDeque::<Da::FilteredBlock>::new();
+        let pending_l1 = &mut pending_l1_blocks;
+        loop {
+            select! {
+                _ = &mut l1_handle => {panic!("l1 sync handle exited unexpectedly");},
+                _ = &mut l2_handle => {panic!("l2 sync handle exited unexpectedly");},
+                Some(l1_block) = l1_rx.recv() => {
+                    pending_l1.push_back(l1_block);
```

Comment:
@eyusufatik @kpp do you guys think this is a potential race condition?

1 worker is pushing to the list and another is processing until this list is empty.

However, every 1 second we are spawning a new task to process the list which means that we might have multiple tasks processing the same items in the list.

Reply:
`l1_sync` worker cannot push before the list is processed i.e. before the `process_l1_block_and_generate_proof` function returns

We are not spawning a new task, `select!` macro itself is one task